src/Lean/Meta/ArgsPacker.lean

```diff
@@ -0,0 +1,534 @@
+    if i >= numFuncs - 1 then
+      return arg
+    else
+      (← whnfD type).withApp fun f args => do
+        assert! args.size == 2
+        if i == fidx then
+          return mkApp3 (mkConst ``PSum.inl f.constLevels!) args[0]! args[1]! arg
+        else
+          let r ← go (i+1) args[1]!
+          return mkApp3 (mkConst ``PSum.inr f.constLevels!) args[0]! args[1]! r
+    termination_by numFuncs - 1 - i
+  go 0 domain
+
+/--
+Unpacks a mutually packed argument created with `Mutual.mk` returning the
+argument and function index.
+
+Throws an error if the expression is not of that form.
+-/
+def unpack (numFuncs : Nat) (expr : Expr) : MetaM (Nat × Expr) := do
+  let mut funidx := 0
+  let mut e := expr
+  while funidx + 1 < numFuncs do
+    if e.isAppOfArity ``PSum.inr 3 then
+      e := e.getArg! 2
+      funidx := funidx + 1
+    else if e.isAppOfArity ``PSum.inl 3 then
+      e := e.getArg! 2
+      break
+    else
+      throwError "Unexpected expression while unpacking mutual argument:{indentExpr expr}"
+  return (funidx, e)
+
+
+/--
+Given unary types `(x : Aᵢ) → Rᵢ[x]`, and `(x : A₁ ⊕ A₂ …)`, calculate the packed codomain
+```
+(match x with | inl x₁ => R₁[x₁] | inr x₂ => R₂[x₂] | …
+```
+This function assumes (and does not check) that `Rᵢ` all have the same level.
+
+If all codomains are equal and non-dependent, just returns that.
+-/
+def mkCodomain (types : Array Expr) (x : Expr) : MetaM Expr := do
+  let types ← types.mapM whnfForall
+  if allEqual (types.map (·.bindingBody!)) then
+    return types[0]!.bindingBody!
+
+  let u ← forallBoundedTelescope types[0]! (some 1) fun _ body => getLevel body
+  let rec go (x : Expr) (i : Nat) : MetaM Expr := do
+    if i < types.size - 1 then
+      let xType ← whnfD (← inferType x)
+      assert! xType.isAppOfArity ``PSum 2
+      let xTypeArgs := xType.getAppArgs
+      let casesOn := mkConst ``PSum.casesOn (mkLevelSucc u :: xType.getAppFn.constLevels!)
+      let casesOn := mkAppN casesOn xTypeArgs -- parameters
+      let casesOn := mkApp casesOn (← mkLambdaFVars #[x] (mkSort u)) -- motive
+      let casesOn := mkApp casesOn x -- major
+      let minor1 ← withLocalDeclD (← mkFreshUserName `_x) xTypeArgs[0]! fun x => do
+        mkLambdaFVars #[x] (types[i]!.bindingBody!.instantiate1 x)
+      let minor2 ← withLocalDeclD (← mkFreshUserName `_x) xTypeArgs[1]! fun x => do
+        mkLambdaFVars #[x] (← go x (i+1))
+      return mkApp2 casesOn minor1 minor2
+    else
+      return types[i]!.bindingBody!.instantiate1 x
+    termination_by types.size - 1 - i
+  go x 0
+  where
+    allEqual (codomains : Array Expr) : Bool :=
+      codomains.all (! ·.hasLooseBVars) && codomains.pop.all (· == codomains.back)
+
+/-
+Given types `(x : A) → R₁[x]` and `(z : B) → R₂[z]`, returns the type
+```
+(x : A ⊕' B) → (match x with | .inl x => R₁[x] | .inr R₂[z]
+```
+if the codomains are dependent, or
+```
+(x : A ⊕' B) → R
+```
+if they are all the same.
+
+-/
```

Comment:
TODO: This heuristic isn't great. The caller knows whether they require a dependent or a non-depenent arrow, so this should be two functions, or an option.